File: Day17/Java/Main.java
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.util.List;

public class Main {

	public static void main(String[] args) throws IOException {
		File input = new File("input").getAbsoluteFile();
		if (!input.exists()) {
			System.out.println(input.getPath() + " does not exist!");
			System.out.println("trying " + input.getPath() + ".txt instead.");
			input = new File(input.getPath() + ".txt");
			if (!input.exists()) {
				System.err.println("neither " + input.getPath().substring(0, input.getPath().length() - 4) + ",");
				System.err.println("nor " + input.getPath() + " exists!");
				return;
			}
		}

		final int cycles = 6;
		List<String> lines = Files.readAllLines(input.toPath());
		boolean[][][] dimension1 = new boolean[cycles * 2 + 1][cycles * 2 + lines.size()][cycles * 2
				+ lines.get(0).length()];
		boolean[][][] dimension2 = new boolean[cycles * 2 + 1][cycles * 2 + lines.size()][cycles * 2
				+ lines.get(0).length()];
		boolean[][][][] dimension3 = new boolean[cycles * 2 + 1][cycles * 2 + 1][cycles * 2 + lines.size()][cycles * 2
				+ lines.get(0).length()];
		boolean[][][][] dimension4 = new boolean[cycles * 2 + 1][cycles * 2 + 1][cycles * 2 + lines.size()][cycles * 2
				+ lines.get(0).length()];
		for (int y = 0; y < lines.size(); y++) {
			String line = lines.get(y);
			for (int x = 0; x < line.length(); x++) {
				dimension1[cycles][cycles + y][cycles + x] = line.charAt(x) == '#';
				dimension2[cycles][cycles + y][cycles + x] = line.charAt(x) == '#';
				dimension3[cycles][cycles][cycles + y][cycles + x] = line.charAt(x) == '#';
				dimension4[cycles][cycles][cycles + y][cycles + x] = line.charAt(x) == '#';
			}
		}

		for (int i = 0; i < cycles; i++) {
			if (i % 2 == 0) {
				simulate3D(dimension1, dimension2);
			} else {
				simulate3D(dimension2, dimension1);
			}
		}

		int active = 0;
		for (int z = 0; z < dimension1.length; z++) {
			for (int y = 0; y < dimension1[z].length; y++) {
				for (int x = 0; x < dimension1[z][y].length; x++) {
					if (dimension2[z][y][x]) {
						active++;
					}
				}
			}
		}

		System.out.println("Found " + active + " active cubes in 3D space.");

		for (int i = 0; i < cycles; i++) {
			if (i % 2 == 0) {
				simulate4D(dimension3, dimension4);
			} else {
				simulate4D(dimension4, dimension3);
			}
		}

		active = 0;
		for (int w = 0; w < dimension3.length; w++) {
			for (int z = 0; z < dimension3[w].length; z++) {
				for (int y = 0; y < dimension3[w][z].length; y++) {
					for (int x = 0; x < dimension3[w][z][y].length; x++) {
						if (dimension4[w][z][y][x]) {
							active++;
						}
					}
				}
			}
		}

		System.out.println("Found " + active + " active cubes in 4D space.");
	}

	private static void simulate3D(boolean[][][] current, boolean[][][] last) {
		for (int z = 0; z < current.length; z++) {
			for (int y = 0; y < current[z].length; y++) {
				for (int x = 0; x < current[z][y].length; x++) {
					current[z][y][x] = check3D(last, x, y, z);
				}
			}
		}
	}

	private static boolean check3D(boolean[][][] last, int x, int y, int z) {
		int activeNeighbors = 0;
		for (int i = -1; i < 2; i++) {
			for (int j = -1; j < 2; j++) {
				for (int k = -1; k < 2; k++) {
					if (z + i < 0 || z + i >= last.length || y + j < 0 || y + j >= last[z + i].length || x + k < 0
							|| x + k >= last[z + i][y + j].length) {
						continue;
					} else if (i == 0 && j == 0 && k == 0) {
						continue;
					} else if (last[z + i][y + j][x + k]) {
						activeNeighbors++;
						if (activeNeighbors > 3) {
							return false;
						}
					}
				}
			}
		}

		if (last[z][y][x]) {
			return activeNeighbors == 2 || activeNeighbors == 3;
		} else {
			return activeNeighbors == 3;
		}
	}

	private static void simulate4D(boolean[][][][] current, boolean[][][][] last) {
		for (int w = 0; w < current.length; w++) {
			for (int z = 0; z < current[w].length; z++) {
				for (int y = 0; y < current[w][z].length; y++) {
					for (int x = 0; x < current[w][z][y].length; x++) {
						current[w][z][y][x] = check4D(last, w, x, y, z);
					}
				}
			}
		}
	}

	private static boolean check4D(boolean[][][][] last, int w, int x, int y, int z) {
		int activeNeighbors = 0;
		for (int i = -1; i < 2; i++) {
			for (int j = -1; j < 2; j++) {
				for (int k = -1; k < 2; k++) {
					for (int l = -1; l < 2; l++) {
						if (w + i < 0 || w + i >= last.length || z + j < 0 || z + j >= last[w + i].length) {
							continue;
						} else if (y + k < 0 || y + k >= last[w + i][z + j].length || x + l < 0
								|| x + l >= last[w + i][z + j][y + k].length) {
							continue;
						} else if (i == 0 && j == 0 && k == 0 && l == 0) {
							continue;
						} else if (last[w + i][z + j][y + k][x + l]) {
							activeNeighbors++;
							if (activeNeighbors > 3) {
								return false;
							}
						}
					}
				}
			}
		}

		if (last[w][z][y][x]) {
			return activeNeighbors == 2 || activeNeighbors == 3;
		} else {
			return activeNeighbors == 3;
		}
	}
}

File: Day17/Python/Main.py
import os

def main():
    input = os.path.dirname(os.getcwd())
    input = os.path.join(input, "input")
    if not os.path.exists(input):
        print(input + " does not exist!")
        input = os.path.join(os.path.dirname(input), "input.txt")
        print(f"trying {input} instead.")
        if not os.path.exists(input):
            print(input + " does not exist!")
            return;

    cycles = 6
    lines = open(input, 'r').readlines()
    dimension1 = [[[False]*(cycles * 2 + len(lines[0]) - 1) for j in range(cycles * 2 + len(lines))] for i in range(cycles * 2 + 1)]
    dimension2 = [[[False]*(cycles * 2 + len(lines[0]) - 1) for j in range(cycles * 2 + len(lines))] for i in range(cycles * 2 + 1)]
    dimension3 = [[[[False]*(cycles * 2 + len(lines[0]) - 1) for k in range(cycles * 2 + len(lines))] for j in range(cycles * 2 + 1)] for i in range(cycles * 2 + 1)]
    dimension4 = [[[[False]*(cycles * 2 + len(lines[0]) - 1) for k in range(cycles * 2 + len(lines))] for j in range(cycles * 2 + 1)] for i in range(cycles * 2 + 1)]
    for y in range(len(lines)):
        line = lines[y]
        for x in range(len(lines[y]) - 1):
            dimension1[cycles][cycles + y][cycles + x] = dimension2[cycles][cycles + y][cycles + x] = lines[y][x] == '#'
            dimension3[cycles][cycles][cycles + y][cycles + x] = dimension4[cycles][cycles][cycles + y][cycles + x] = lines[y][x] == '#'

    for i in range(cycles):
        if i % 2 == 0:
            simulate3D(dimension1, dimension2)
        else:
            simulate3D(dimension2, dimension1)

    active = 0
    for z in range(len(dimension1)):
        for y in range(len(dimension1[z])):
            for x in range(len(dimension1[z][y])):
                if cycles % 2 == 0 and dimension2[z][y][x]:
                    active += 1
                elif cycles % 2 == 1 and dimension1[z][y][x]:
                    active += 1

    print(f"Found {active} active cubes in 3D space.")

    for i in range(cycles):
        if i % 2 == 0:
            simulate4D(dimension3, dimension4)
        else:
            simulate4D(dimension4, dimension3)

    active = 0
    for w in range(len(dimension3)):
        for z in range(len(dimension3[w])):
            for y in range(len(dimension3[w][z])):
                for x in range(len(dimension3[w][z][y])):
                    if cycles % 2 == 0 and dimension4[w][z][y][x]:
                        active += 1
                    elif cycles % 2 == 1 and dimension3[w][z][y][x]:
                        active += 1

    print(f"Found {active} active cubes in 4D space.")

def simulate3D(current, last):
    for z in range(len(current)):
        for y in range(len(current[z])):
            for x in range(len(current[z][y])):
                current[z][y][x] = check3D(last, x, y, z)

def check3D(last, x, y, z):
    activeNeighbors = 0
    for i in range(-1, 2):
        for j in range(-1, 2):
            for k in range(-1, 2):
                if z + i < 0 or z + i >= len(last) or y + j < 0 or y + j >= len(last[z + i]) or x + k < 0 or x + k >= len(last[z + i][y + j]):
                    continue
                elif i == 0 and j == 0 and k == 0:
                    continue
                elif last[z + i][y + j][x + k]:
                    activeNeighbors += 1
                    if activeNeighbors > 3:
                        return False

    if last[z][y][x]:
        return activeNeighbors == 2 or activeNeighbors == 3
    else:
        return activeNeighbors == 3

def simulate4D(current, last):
    for w in range(len(current)):
        for z in range(len(current[w])):
            for y in range(len(current[w][z])):
                for x in range(len(current[w][z][y])):
                    current[w][z][y][x] = check4D(last, w, x, y, z)

def check4D(last, w, x, y, z):
    activeNeighbors = 0
    for i in range(-1, 2):
        for j in range(-1, 2):
            for k in range(-1, 2):
                for l in range(-1, 2):
                    if w + i < 0 or w + i >= len(last) or z + j < 0 or z + j >= len(last[w + i]):
                        continue
                    elif y + k < 0 or y + k >= len(last[w + i][z + j]) or x + l < 0 or x + l >= len(last[w + i][z + j][y + k]):
                        continue
                    elif i == 0 and j == 0 and k == 0 and l == 0:
                        continue
                    elif last[w + i][z + j][y + k][x + l]:
                        activeNeighbors += 1
                        if activeNeighbors > 3:
                            return False

    if last[w][z][y][x]:
        return activeNeighbors == 2 or activeNeighbors == 3
    else:
        return activeNeighbors == 3

if __name__ == '__main__':
    main()
